Testing Content Update (#150)
JoshOrndorff/recipes: text/testing/mock.md

# Mock Runtime for Unit Testing

> *see [root](./index.md) for list of kitchen pallets with unit test coverage*

Reviewer: I know this line existed prior to the PR, but is it still relevant? If not can you please remove it?

Author: I see not all pallets have unit test yet. So I think it is good to list out pallets with test, and readers could quickly refer back to these tests.

There are two main patterns on writing tests for pallets. We can put the tests:

1. At the bottom of the pallet, place unit tests in a separate Rust module with a special compilation flag

	```rust, ignore
	#[cfg(test)]
	mod tests {
		// -- snip --
	}
	```

2. In a separate file called `tests.rs` inside `src` folder, and conditionally include tests inside the main `lib.rs`. At the top of the `lib.rs`

	```rust, ignore
	#[cfg(test)]
	mod tests;
	```

Now, to use the logic from the pallet under test, bring `Module` and `Trait` into scope.

```rust, ignore
use crate::{Module, Trait};
```

## Create the Outer Environment for Mock Runtime

Before we create the mock runtime that take our pallet to run tests, we first need to create the
outer environment for the runtime as follows:

```rust, ignore
use support::{impl_outer_event, impl_outer_origin, parameter_types};
use runtime_primitives::{Perbill, traits::{IdentityLookup, BlakeTwo256}, testing::Header};
use runtime_io;
use primitives::{H256};

// We define the outer `Origin` enum and `Event` enum.

> // You may not aware these enums are created when writing the runtime/pallet, it is

Reviewer: ```suggestion
// You may not be aware that these enums are created when writing the runtime/pallet; it is
```

Next, we create a new type that wraps the mock `TestRuntime` in the pallet's `Module`.

```rust, ignore
pub type TestPallet = Module<TestRuntime>;
```

It may be helpful to read this as type aliasing our configured mock runtime to work with the pallet's `Module`, which is what is ultimately being tested.

## `impl system::Trait`

In many cases, the pallet's `Trait` inherits `system::Trait` like:

```rust, ignore
pub trait Trait: system::Trait {
	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
}
```

Before we can create a mock runtime that take our pallet to run tests, we first need to create the

> outer environment for the runtime, as follows:

Reviewer: The information you've added here is good and should be kept. But it is added right in the middle of another thought. You should finish showing how to `impl system::Trait` then explain how to impl the outer enums.

Author: I put this on top of this section.

```rust, ignore
use support::{impl_outer_event, impl_outer_origin, parameter_types};
use runtime_primitives::{Perbill, traits::{IdentityLookup, BlakeTwo256}, testing::Header};
use runtime_io;
use primitives::{H256};

// We define the outer `Origin` enum, and `Event` enums.
// You may not aware these enums are created when writing the pallet, it is because

> //   they are created through the macro `decl_event`, `decl_module` that we use.

Reviewer: The _outer_ enums like `outer_origin`, `outer_event`, `outer_error` etc are created by `construct_runtime!`.

It's the inner enums that are constructed by macros like `decl_event!`.

Author: Indeed. Just checked with `cargo expand`.

Next, we create a new type that wraps the mock `TestRuntime` in the pallet's `Module`.

```rust, ignore
pub type TestPallet = Module<TestRuntime>;
```

It may be helpful to read this as type aliasing our configured mock runtime to work with the pallet's `Module`, which is what is ultimately being tested.

In many cases, the pallet's `Trait` is further bound by `system::Trait` like:

```rust, ignore
pub trait Trait: system::Trait {
	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
}
```

The mock runtime must inherit and define the `system::Trait` associated types
(*[remember](https://substrate.dev/recipes/traits/index.html)*). To do so, `impl` the `system::Trait`
for `TestRuntime` with types created previously and imported from other crates.

```rust, ignore
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct TestRuntime;

parameter_types! {
	pub const BlockHashCount: u64 = 250;
	pub const MaximumBlockWeight: u32 = 1024;
	pub const MaximumBlockLength: u32 = 2 * 1024;
	pub const AvailableBlockRatio: Perbill = Perbill::one();
}

> // First, implement the system pallet trait for test runtime

Reviewer: ```suggestion
// First, implement the system pallet's configuration trait for `TestRuntime`
```

impl system::Trait for TestRuntime {
	type Origin = Origin;
	type Index = u64;
	type Call = ();
	type BlockNumber = u64;
	type Hash = H256;
	type Hashing = BlakeTwo256;
	type AccountId = u64;
	type Lookup = IdentityLookup<Self::AccountId>;
	type Header = Header;
	// To test events, use `TestEvent`. Otherwise, use the commented line
	type Event = TestEvent;
	// type Event = ();
	type BlockHashCount = BlockHashCount;
	type MaximumBlockWeight = MaximumBlockWeight;
	type MaximumBlockLength = MaximumBlockLength;
	type AvailableBlockRatio = AvailableBlockRatio;
	type Version = ();
	type ModuleToIndex = ();
}

> // Then implement our own pallet trait for test runtime

Reviewer: ```suggestion
// Then implement our own pallet's configuration trait for `TestRuntime`
```
